pallets/ajuna-tournament/src/lib.rs

```diff
@@ -0,0 +1,753 @@
+				T::Currency::transfer(
+					creator,
+					&treasury_account,
+					reward,
+					ExistenceRequirement::KeepAlive,
+				)?;
+			}
+
+			if let GoldenDuckConfig::Enabled(percentage) = config.golden_duck_config {
+				GoldenDucks::<T, I>::insert(
+					season_id,
+					next_tournament_id,
+					GoldenDuckStateFor::<T, I>::Enabled(percentage, None),
+				);
+			}
+
+			Tournaments::<T, I>::insert(season_id, next_tournament_id, config);
+
+			Self::deposit_event(Event::<T, I>::TournamentCreated {
+				season_id: *season_id,
+				tournament_id: next_tournament_id,
+			});
+
+			Ok(next_tournament_id)
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> TournamentRanker<T::SeasonId, T::RankedEntity, T::EntityId>
+		for Pallet<T, I>
+	{
+		fn try_rank_entity_in_tournament_for<R>(
+			season_id: &T::SeasonId,
+			entity_id: &T::EntityId,
+			entity: &T::RankedEntity,
+			ranker: &R,
+		) -> DispatchResult
+		where
+			R: EntityRank<EntityId = T::EntityId, Entity = T::RankedEntity>,
+		{
+			let tournament_id = Self::try_get_active_tournament_id_for(season_id)?;
+			let tournament_config = Self::get_active_tournament_config_for(season_id)
+				.ok_or::<Error<T, I>>(Error::<T, I>::TournamentNotFound)?;
+
+			TournamentRankings::<T, I>::mutate(season_id, tournament_id, |table| {
+				if let Err(index) = table.binary_search_by(|(other_id, other)| {
+					ranker.rank_against((entity_id, entity), (other_id, other))
+				}) {
+					match Self::try_update_rank_table(
+						table,
+						&tournament_config,
+						index,
+						entity_id,
+						entity,
+					) {
+						None => Ok(()),
+						Some(true) => {
+							Self::deposit_event(
+								crate::pallet::Event::<T, I>::EntityEnteredRanking {
+									season_id: *season_id,
+									tournament_id,
+									entity_id: entity_id.clone(),
+								},
+							);
+							Ok(())
+						},
+						Some(false) => Err(Error::<T, I>::FailedToRankEntity.into()),
+					}
+				} else {
+					Ok(())
+				}
+			})
+		}
+
+		fn try_rank_entity_for_golden_duck(
+			season_id: &T::SeasonId,
+			entity_id: &T::EntityId,
+		) -> DispatchResult {
+			let tournament_id = Self::try_get_active_tournament_id_for(season_id)?;
+
+			GoldenDucks::<T, I>::mutate(season_id, tournament_id, |state| match state {
+				GoldenDuckState::Enabled(payout_perc, None) => {
+					*state = GoldenDuckState::Enabled(*payout_perc, Some(entity_id.clone()));
+
+					Self::deposit_event(crate::pallet::Event::<T, I>::EntityBecameGoldenDuck {
+						season_id: *season_id,
+						tournament_id,
+						entity_id: entity_id.clone(),
+					});
+				},
+				GoldenDuckState::Enabled(payout_perc, Some(ref entry_id))
+					if entity_id < entry_id =>
+				{
+					*state = GoldenDuckState::Enabled(*payout_perc, Some(entity_id.clone()));
+
+					Self::deposit_event(crate::pallet::Event::<T, I>::EntityBecameGoldenDuck {
+						season_id: *season_id,
+						tournament_id,
+						entity_id: entity_id.clone(),
+					});
+				},
+				_ => {},
+			});
```

Comment:
Reduce duplicate code and make more readable:


```rust
GoldenDuckState::Enabled(payout_perc, maybe_entry_id) => {
			if  maybe_entry_id.map(|e| entity_id < e).unwrap_or(true) =>
		{
			*state = GoldenDuckState::Enabled(*payout_perc, Some(entity_id.clone()));

			Self::deposit_event(crate::pallet::Event::<T, I>::EntityBecameGoldenDuck {
				season_id: *season_id,
				tournament_id,
				entity_id: entity_id.clone(),
			});
		},

```

Reply:
not 100% sure if it is more understandable, now that I read my code XD

Reply:
I tired to reduce duplication as much as I could with the new commits, please take a look!

Reply:
Yeah, the issue is that the data structure makes the code handling difficult here, we mix a setting `Disabled`/`Enabled` here with the state (maybe winner). We can make the code definitely more ergonomic by splitting the state into two separate things to correctly have separation of concerns of config & state. What was your reasoning for putting it into the same storage? Reduce db-reads? I think would only optimize on that level if we run into performance bottlenecks if it makes the code more complicated, but I guess this is opinionated, and I can see where you are coming from.

A good compromise here might be to hide the internal complexity in a setter/getter:


```rust
impl GoldenDuckState {
   pub fn winner() -> Option<EntityId>;

   // updates the winner if enabled and entity_id > entry_id
   pub fn maybe_update_winner()
}

```

Reply:
If we mix it in one type instead of two than I would suggest to at least separate state and setting within the type aka:

```rust
struct GoldenDuckState {
   pub config: GoldenDuckConfig,
   // contains a potential winner 
   winner: Option<EntityId>,
}

impl GoldenDuckState {
   pub fn new(config: GoldenDuckConfig) -> Self {
      Self { config, winner: None }
   }

   pub fn set_winner(&mut self, winner: EntityId) -> Result<(), ()> {
      match self.config {
          Enabled => {
             self.winner = Some(winner);
             Ok(())
          },
          Disabled => Err(())
     }
   }

   pub fn winner() -> Option<EntityId>
}

```